apps/authentication/views.py

```diff
@@ -83,9 +85,12 @@ def register(request):
 
                 # Create the registration token
                 token = uuid.uuid4().hex
-                rt = RegisterToken(user=user, email=email.email, token=token)
-                rt.save()
-
+                try:
+                    rt = RegisterToken(user=user, email=email.email, token=token)
+                    rt.save()
+                    log.info('Successfully registered token for %s' % request.user)
+                except:
+                    log.error('Failed to register token for %s' % request.user)
```

Comment:
Yes. No, not at all. This is a bit of a weird case. 

By overriding the `create()`-method, we'll override the `create()`-method for _all_ RegisterToken objects. We'll also make sure that it behaves just like it does today, but move the generation of a unique token to the `create()` method instead of manually generating it first, and then adding it to the model.

Yes, we have to keep generating codes until we find a string that works. However, the [probability of collision](https://en.wikipedia.org/wiki/Universally_unique_identifier#Random_UUID_probability_of_duplicates) is very small, though it exists. This means we most likely won't ever need to re-generate a code, and if we do, we'll most likely not have to do it a lot of times.

Here are some of my thoughts in some kind of code.

```
registertoken = RegisterToken.objects.create(user=user, email=email)

# Django will now create an object for user and email, 
# which will automagically generate a token on the object like so:

>>> registertoken.token 
a494f4101ad54f2bb91b82612bf6320d  # uuid.uuid4().hex
```

This way we don't need to remember to check for integrity errors every time we create a token.

Or, we can do it like this
```
def generate_unique_token():
    token = uuid.uuid4().hex
    if RegisterToken.objects.filter(token=token):
        token = generate_unique_token()
    else:
        return token

...

token = generate_unique_token()
RegisterToken.objects.create(user=user, email=email, token=token)
```
or, as you suggested, a loop:
```
token = uuid.uuid4().hex
while RegisterToken.objects.filter(token=token):
    token = uuid.uuid4().hex

RegisterToken.objects.create(user=user, email=email, token=token)
```

The first way I suggested would implement either of these two checks, or some other, to make sure a unique token is generated. This could be implemented by hooking onto a pre_save signal (an event which is fired _before_ a model is persisted in the database).

In my opinion, the first option is the cleanest. It will place the token validation in one place, and developers won't have to think about verifying if the token already exists, Django will take care of that in the future.